From 49460b55aefa531401a242f9ba3b25c7443ee344 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 12 Feb 2025 14:43:45 +0800
Subject: [PATCH 088/104] mtk: mt76: mt7990: add mt7990 background radar
 capability eeprom check

According to SA & SE from the Logan team, mt7990 (whether i/efem) can
switch from 3T3R (no ZWDFS) to 2T2R + 1R (ZWDFS) without hardware reworking.
In logan, when the eeprom specifies 3T3R, then background radar capability is disabled to prevent conflicts.
Therefore, align it in mt76.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/debugfs.c |  2 +-
 mt7996/eeprom.c  | 28 ++++++++++++++++++++++++++++
 mt7996/init.c    |  2 +-
 mt7996/mt7996.h  | 23 +----------------------
 4 files changed, 31 insertions(+), 24 deletions(-)

diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
index 6ae8bf2a..a875253c 100644
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -283,7 +283,7 @@ mt7996_rdd_monitor(struct seq_file *s, void *data)
 
 	mutex_lock(&dev->mt76.mutex);
 
-	if (!mt7996_has_background_radar(dev)) {
+	if (!mt7996_eeprom_has_background_radar(dev)) {
 		seq_puts(s, "no background radar capability\n");
 		goto out;
 	}
diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 079fb4b3..cfaa760f 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -649,6 +649,34 @@ s8 mt7996_eeprom_get_power_delta(struct mt7996_dev *dev, int band)
 	return val & MT_EE_RATE_DELTA_SIGN ? delta : -delta;
 }
 
+bool mt7996_eeprom_has_background_radar(struct mt7996_dev *dev)
+{
+	switch (mt76_chip(&dev->mt76)) {
+	case MT7996_DEVICE_ID:
+		if (dev->var.type == MT7996_VAR_TYPE_233)
+			return false;
+		break;
+	case MT7992_DEVICE_ID:
+		if (dev->var.type == MT7992_VAR_TYPE_23 ||
+		    dev->var.type == MT7992_VAR_TYPE_24)
+			return false;
+		break;
+	case MT7990_DEVICE_ID: {
+		u8 path, rx_path, nss, *eeprom = dev->mt76.eeprom.data;
+
+		mt7996_eeprom_parse_stream(eeprom, MT_BAND1, &path, &rx_path, &nss);
+		/* Disable background radar capability in 3T3R */
+		if (path == 3 || rx_path == 3)
+			return false;
+		break;
+	}
+	default:
+		return false;
+	}
+
+	return true;
+}
+
 const u8 mt7996_sku_group_len[] = {
 	[SKU_CCK] = 4,
 	[SKU_OFDM] = 8,
diff --git a/mt7996/init.c b/mt7996/init.c
index 7e68ffb0..922fb667 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -560,7 +560,7 @@ mt7996_init_wiphy(struct ieee80211_hw *hw, struct mtk_wed_device *wed)
 
 	wiphy_ext_feature_set(wiphy, NL80211_EXT_FEATURE_OPERATING_CHANNEL_VALIDATION);
 	wiphy_ext_feature_set(wiphy, NL80211_EXT_FEATURE_BEACON_PROTECTION);
-	if (mt7996_has_background_radar(dev) &&
+	if (mt7996_eeprom_has_background_radar(dev) &&
 	    (!mdev->dev->of_node ||
 	     !of_property_read_bool(mdev->dev->of_node,
 				    "mediatek,disable-radar-background")))
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index ac380337..d39fd396 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1046,28 +1046,6 @@ mt7996_radio_valid(struct mt7996_dev *dev, u8 radio)
 	return true;
 }
 
-static inline bool
-mt7996_has_background_radar(struct mt7996_dev *dev)
-{
-	switch (mt76_chip(&dev->mt76)) {
-	case MT7996_DEVICE_ID:
-		if (dev->var.type == MT7996_VAR_TYPE_233)
-			return false;
-		break;
-	case MT7992_DEVICE_ID:
-		if (dev->var.type == MT7992_VAR_TYPE_23 ||
-		    dev->var.type == MT7992_VAR_TYPE_24)
-			return false;
-		break;
-	case MT7990_DEVICE_ID:
-		break;
-	default:
-		return false;
-	}
-
-	return true;
-}
-
 static inline bool
 mt7996_has_ext_eeprom(struct mt7996_dev *dev)
 {
@@ -1246,6 +1224,7 @@ int mt7996_eeprom_parse_hw_cap(struct mt7996_dev *dev, struct mt7996_phy *phy);
 int mt7996_eeprom_get_target_power(struct mt7996_dev *dev,
 				   struct ieee80211_channel *chan);
 s8 mt7996_eeprom_get_power_delta(struct mt7996_dev *dev, int band);
+bool mt7996_eeprom_has_background_radar(struct mt7996_dev *dev);
 int mt7996_dma_init(struct mt7996_dev *dev);
 void mt7996_dma_reset(struct mt7996_dev *dev, bool force);
 void mt7996_dma_prefetch(struct mt7996_dev *dev);
-- 
2.45.2

